tools/ami-cleanup/src/internal/cleanup.go

```diff
@@ -0,0 +1,239 @@
+	totalSnapshotCount := 0
+	for _, devImage := range devImages {
+		imageSpace, err := ai.cleanupImageIfOld(ctx, ec2Client, devImage)
+		if err != nil {
+			return 0, 0, trace.Wrap(err, "failed to cleanup dev image %q", devImage.Name)
+		}
+
+		if imageSpace == 0 {
+			continue
+		}
+
+		totalSpaceRecovered += imageSpace
+		totalImagesDeleted++
+		totalSnapshotCount += len(devImage.BlockDeviceMappings)
+	}
+
+	return totalSpaceRecovered, totalImagesDeleted, nil
+}
+
+func (ai *ApplicationInstance) getEnabledRegions(ctx context.Context, client IAccountApi) ([]accountTypes.Region, error) {
+	return getAllWithPagination(
+		func(previousToken *string) (*string, []accountTypes.Region, error) {
+			results, err := client.ListRegions(ctx, &account.ListRegionsInput{
+				RegionOptStatusContains: []accountTypes.RegionOptStatus{
+					accountTypes.RegionOptStatusEnabled,
+					accountTypes.RegionOptStatusEnabledByDefault,
+				},
+				NextToken: previousToken,
+			})
+			if err != nil {
+				return nil, nil, trace.Wrap(err, "failed to request enabled regions")
+			}
+
+			return results.NextToken, results.Regions, nil
+		},
+	)
+}
+
+func (ai *ApplicationInstance) getDevImagesInRegion(ctx context.Context, client IEc2Api) ([]ec2Types.Image, error) {
+	// These are a weird language workaround for getting a pointer to a boolean literal
+	includeTrue := true
+	nameFilterName := "name"
+	stateFilterName := "state"
+	requestInput := &ec2.DescribeImagesInput{
+		Filters: []ec2Types.Filter{
+			{
+				Name:   &nameFilterName,
+				Values: []string{"*dev*"},
+			},
+			{
+				Name:   &stateFilterName,
+				Values: []string{string(ec2Types.ImageStateAvailable)},
+			},
+		},
+		IncludeDeprecated: &includeTrue,
+		IncludeDisabled:   &includeTrue,
+		Owners:            []string{"self"},
+	}
+
+	return getAllWithPagination(
+		func(previousToken *string) (*string, []ec2Types.Image, error) {
+
+			requestInput.NextToken = previousToken
+			results, err := client.DescribeImages(ctx, requestInput)
+			if err != nil {
+				return nil, nil, trace.Wrap(err, "failed to request a dev images")
+			}
+
+			return results.NextToken, results.Images, nil
+		},
+	)
+}
+
+// Repeatedly calls `action` until the returned `nextToken` is null, and accumulates the results.
+// Useful for AWS API calls who's results may be paginated.
+func getAllWithPagination[T any](action func(previousToken *string) (nextToken *string, results []T, err error)) ([]T, error) {
+	var previousToken *string
+	var results []T
+	for {
+		nextToken, newResults, err := action(previousToken)
+		if err != nil {
+			return results, trace.Wrap(err, "failed to get the next set of results")
+		}
+
+		results = append(results, newResults...)
+		if nextToken == nil {
+			return results, nil
+		}
+		previousToken = nextToken
+	}
+}
+
+func (ai *ApplicationInstance) cleanupImageIfOld(ctx context.Context, client IEc2Api, image ec2Types.Image) (int32, error) {
+	creationDate, err := iso8601.ParseString(*image.CreationDate)
```

Comment:
Do we need a 3rd party dependency for this? Can `time.Parse` do this for you?

Reply:
Strictly speaking, yes. `time` does not fully support iso8601, just a subset. According to the docs the returned value is an iso8601 timestamp string.